File: python/examples/collection-sliding-window_test.py
"""カード collection-sliding-window の Contract を検証するテスト"""

from itertools import count

import pytest
from more_itertools import sliding_window, windowed


def test_windows_keep_order_and_advance_by_step():
    """窓は先頭から step 刻みで並び、窓の中も元の並び順"""
    assert list(windowed([1, 2, 3, 4, 5], 3)) == [(1, 2, 3), (2, 3, 4), (3, 4, 5)]
    assert list(windowed([1, 2, 3, 4, 5], 3, step=2)) == [(1, 2, 3), (3, 4, 5)]


def test_does_not_mutate_input_and_keeps_references():
    """入力を変更せず、各窓は新しいタプルで要素は同じ参照"""
    a = {"k": 1}
    src = [a, {"k": 2}]
    result = list(windowed(src, 2))
    assert src == [{"k": 1}, {"k": 2}]
    assert isinstance(result[0], tuple)
    assert result[0][0] is a


def test_is_lazy_and_single_pass():
    """窓 1 つ分ずつ入力を読み、2 回目の走査は空になる"""
    assert next(windowed(count(), 3)) == (0, 1, 2)
    once = windowed([1, 2, 3], 2)
    assert list(once) == [(1, 2), (2, 3)]
    assert list(once) == []


def test_step_larger_than_n_skips_elements():
    """step > n なら窓の間の要素は飛ばされる"""
    assert list(windowed([1, 2, 3, 4, 5, 6], 2, step=3)) == [(1, 2), (4, 5)]


def test_tail_is_filled_not_dropped():
    """はみ出す末尾は fillvalue で埋める。前の窓に含まれた要素しか残らなければ出力しない"""
    assert list(windowed([1, 2, 3, 4, 5, 6, 7], 3, step=3)) == [(1, 2, 3), (4, 5, 6), (7, None, None)]
    assert list(windowed([1, 2, 3, 4, 5, 6], 3, step=2)) == [(1, 2, 3), (3, 4, 5), (5, 6, None)]
    assert list(windowed([1, 2, 3, 4, 5], 3, step=2)) == [(1, 2, 3), (3, 4, 5)]
    assert list(windowed([1, 2, 3, 4, 5], 2, step=4)) == [(1, 2), (5, None)]


def test_short_input_yields_single_filled_window():
    """入力長が n 未満なら fillvalue で埋めた窓を 1 つだけ返す"""
    assert list(windowed([1, 2], 3)) == [(1, 2, None)]
    assert list(windowed([1, 2], 3, fillvalue=0)) == [(1, 2, 0)]


def test_empty_input_yields_nothing():
    """空のイテラブルは何も返さない"""
    assert list(windowed([], 3)) == []


def test_invalid_n_or_step_raises_value_error():
    """n が 1 未満、または step が 1 未満なら ValueError"""
    with pytest.raises(ValueError):
        list(windowed([1, 2], 0))
    with pytest.raises(ValueError):
        list(windowed([1, 2], -1))
    with pytest.raises(ValueError):
        list(windowed([1, 2], 2, step=0))


def test_sliding_window_drops_incomplete_windows():
    """Alternatives: sliding_window は n 個そろった窓だけを返す"""
    assert list(sliding_window([1, 2, 3, 4], 2)) == [(1, 2), (2, 3), (3, 4)]
    assert list(sliding_window([1, 2], 3)) == []

File: typescript/examples/collection-sliding-window.test.ts
// カード collection-sliding-window の Contract を検証するテスト
import { chunk, windowed } from 'es-toolkit';
import { describe, expect, it } from 'vitest';

describe('collection-sliding-window: windowed', () => {
  it('既定では step 1 で、size の窓を順序どおりに返す', () => {
    expect(windowed([1, 2, 3, 4, 5], 3)).toEqual([
      [1, 2, 3],
      [2, 3, 4],
      [3, 4, 5],
    ]);
  });

  it('入力配列を変更せず、各窓は新しい配列で要素は同じ参照', () => {
    const a = { id: 1 };
    const input = [a, { id: 2 }];
    const result = windowed(input, 1);
    expect(input).toHaveLength(2);
    expect(result[0]).not.toBe(input);
    expect(result[0][0]).toBe(a);
  });

  it('step > size なら窓の間の要素を飛ばす', () => {
    expect(windowed([1, 2, 3, 4, 5], 2, 3)).toEqual([
      [1, 2],
      [4, 5],
    ]);
  });

  it('既定では size に満たない末尾の窓を捨て、配列長が size 未満なら空配列', () => {
    expect(windowed([1, 2, 3, 4, 5], 3, 2)).toEqual([
      [1, 2, 3],
      [3, 4, 5],
    ]);
    expect(windowed([1, 2, 3, 4], 3, 2)).toEqual([[1, 2, 3]]);
    expect(windowed([1, 2], 3)).toEqual([]);
  });

  it('partialWindows: true なら末尾の短い窓も返す', () => {
    expect(windowed([1, 2, 3, 4, 5], 3, 2, { partialWindows: true })).toEqual([
      [1, 2, 3],
      [3, 4, 5],
      [5],
    ]);
    expect(windowed([1, 2], 3, 1, { partialWindows: true })).toEqual([[1, 2], [2]]);
  });

  it('step = size かつ partialWindows: true は chunk と同じ結果になる', () => {
    const input = [1, 2, 3, 4, 5];
    expect(windowed(input, 2, 2, { partialWindows: true })).toEqual(chunk(input, 2));
  });

  it('空配列は空配列を返す', () => {
    expect(windowed([], 2)).toEqual([]);
  });

  it('size または step が正の整数でなければ例外を投げる', () => {
    expect(() => windowed([1, 2, 3], 0)).toThrow('Size must be a positive integer.');
    expect(() => windowed([1, 2, 3], -1)).toThrow();
    expect(() => windowed([1, 2, 3], 1.5)).toThrow();
    expect(() => windowed([1, 2, 3], Number.NaN)).toThrow();
    expect(() => windowed([1, 2, 3], 2, 0)).toThrow('Step must be a positive integer.');
    expect(() => windowed([1, 2, 3], 2, -1)).toThrow();
    expect(() => windowed([1, 2, 3], 2, 1.5)).toThrow();
  });
});
